patches/squirrel.mac/feat_add_new_squirrel_mac_bundle_installation_method_behind_flag.patch

```diff
@@ -0,0 +1,128 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Samuel Attard <samuel.r.attard@gmail.com>
+Date: Mon, 28 Mar 2022 02:43:18 -0700
+Subject: feat: add new Squirrel.Mac bundle installation method behind flag
+
+The 'SquirrelMacEnableDirectContentsWrite' user default in your apps defaults suite
+can be used to control this new installation method.  It is designed to remove the
+requirement that the updating process have write access to its parent directory.
+
+With this feature enabled the updating process only needs write access to it's own
+.app bundle folder, not the owning /Applications folder.  This should allow more
+non-admin users to update applications when appropriately granted group based permissions.
+
+E.g. 775 and chown :staff
+
+diff --git a/Squirrel/SQRLInstaller.m b/Squirrel/SQRLInstaller.m
+index 7dd98ddee4ae0f4e01fd7aaa3486083bff7d0da1..c1f328fa8c3689218ef260347cb8f9d30b789efe 100644
+--- a/Squirrel/SQRLInstaller.m
++++ b/Squirrel/SQRLInstaller.m
+@@ -249,6 +249,7 @@ - (RACSignal *)acquireTargetBundleURLForRequest:(SQRLShipItRequest *)request {
+ 		] reduce:^(NSURL *directoryURL, SQRLCodeSignature *codeSignature) {
+ 			NSURL *targetBundleURL = request.targetBundleURL;
+ 			NSURL *newBundleURL = [directoryURL URLByAppendingPathComponent:targetBundleURL.lastPathComponent];
++			[NSFileManager.defaultManager createDirectoryAtURL:newBundleURL withIntermediateDirectories:FALSE attributes:nil error:nil];
+ 
+ 			return [[SQRLInstallerOwnedBundle alloc] initWithOriginalURL:request.targetBundleURL temporaryURL:newBundleURL codeSignature:codeSignature];
+ 		}]
+@@ -481,10 +482,50 @@ - (RACSignal *)installItemToURL:(NSURL *)targetURL fromURL:(NSURL *)sourceURL {
+ 	NSParameterAssert(targetURL != nil);
+ 	NSParameterAssert(sourceURL != nil);
+ 
++	NSLog(@"Moving bundle from %@ to %@", sourceURL, targetURL);
++
++	// If both the sourceURL and the targetURL exist we can try to skip a permissions check
++	// by moving Thing.app/Contents directly.  This allows us to update applications without
++	// permission to write files into the parent directory of Thing.app
++	//
++	// There is no known case where these directories don't exist but in order to handle
++	// edge cases / race conditions we'll handle it anyway.
++	//
++	// This exists check is non-atomic with the rename call below but that's OK
```

Comment:
maybe try the `Contents` move first and then if that fails, move the parent?

Reply:
There is already a built-in retry system in Squirrel.Mac, assuming the one in a million edge case is somehow hit where the `.app` bundle folder is deleted after this exists check, the odds is happens multiple times in a row with the retry system is effectively 0.  To avoid complexity I'd rather rely on the retry than try to smoothly handle the theoretically impossible edge case.